File: golang/two_pointers/square_sorted_array.go
package two_pointers

func makeSquares(arr []int) []int {
	n := len(arr)
	highSqIdx := n - 1
	squares := make([]int, n)

	left, right := 0, n-1

	for left <= right {
		leftSq, rightSq := arr[left]*arr[left], arr[right]*arr[right]

		if leftSq > rightSq {
			squares[highSqIdx] = leftSq
			left++
		} else {
			squares[highSqIdx] = rightSq
			right--
		}

		highSqIdx--
	}

	return squares
}

File: javascript/src/two_pointers/squareSortedArray.js
export function makeSquares(arr) {
    const n = arr.length;
    let highSqIdx = n - 1;
    let squares = Array(n).fill(0);

    let left = 0,
        right = n - 1;

    while (left <= right) {
        const leftSq = arr[left] * arr[left],
            rightSq = arr[right] * arr[right];

        if (leftSq > rightSq) {
            squares[highSqIdx] = leftSq;
            left++;
        } else {
            squares[highSqIdx] = rightSq;
            right--;
        }

        highSqIdx--;
    }

    return squares;
}
